File: LC1552_MaxTheMinDistance/solution.py
from typing import List
class Solution:
    def maxDistance(self, position: List[int], m: int) -> int:
        position.sort()
        def possible(dist):
            prev=position[0]
            count=1
            for p in position[1:]:
                if p - prev >= dist:
                    prev=p
                    count+=1

            return count >= m
        
        low,high,ans=1,position[-1]-position[0],-1
        while low <= high:
            mid=low+(high-low)//2
            if possible(mid):
                ans = mid
                low = mid+1
            else:
                high = mid -1
        return ans

File: LC1552_MaxTheMinDistance/solution.go
package lc1552maxthemindistance

import "sort"

func maxDistance(position []int, m int) int {
	var ans int
	sort.Ints(position)
	n := len(position)
	low, high := 1, position[n-1]-position[0]

	for low <= high {
		mid := low + (high-low)/2
		if func(dist int) bool {
			prev, count := position[0], 1
			for _, curr := range position[1:] {
				if curr-prev >= dist {
					prev = curr
					count++
				}
			}
			return count >= m
		}(mid) {
			ans = mid
			low = mid + 1
		} else {
			high = mid - 1
		}
	}
	return ans
}
